modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+		unfinishedChunks := r.buildUnfinishedChunks(file, hosts, fcidToHPK)
+		ch = append(ch, unfinishedChunks...)
+	}
+	r.mu.Unlock(id)
+
+	// Init the heap.
+	heap.Init(&ch)
+	return &ch
+}
+
+// managedInsertFileIntoChunkHeap will insert all of the chunks of a file into the
+// chunk heap.
+func (r *Renter) managedInsertFileIntoChunkHeap(f *file, ch *chunkHeap, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) {
+	id := r.mu.Lock()
+	unfinishedChunks := r.buildUnfinishedChunks(f, hosts, fcidToHPK)
+	for i := 0; i < len(unfinishedChunks); i++ {
+		heap.Push(ch, unfinishedChunks)
+	}
+	r.mu.Unlock(id)
+}
+
+// managedPrepareNextChunk takes the next chunk from the chunk heap and prepares
+// it for upload. Preparation includes blocking until enough memory is
+// available, fetching the logical data for the chunk (either from the disk or
+// from the network), erasure coding the logical data into the physical data,
+// and then finally passing the work onto the workers.
+func (r *Renter) managedPrepareNextChunk(ch *chunkHeap, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) {
+	// Grab the next chunk, loop until we have enough memory, update the amount
+	// of memory available, and then spin up a thread to asynchronously handle
+	// the rest of the chunk tasks.
+	memoryAvailable := r.managedMemoryAvailableGet()
+	nextChunk := ch.Pop().(*unfinishedChunk)
+	for nextChunk.memoryNeeded > memoryAvailable {
+		select {
+		case newFile := <-r.newUploads:
+			r.managedInsertFileIntoChunkHeap(newFile, ch, hosts, fcidToHPK)
+		case <-r.newMemory:
+		case <-r.tg.StopChan():
+		}
+	}
+	r.managedMemoryAvailableSub(nextChunk.memoryNeeded)
+	go r.managedFetchAndRepairChunk(nextChunk)
+}
+
+// threadedRepairScan is a background thread that checks on the health of files,
+// tracking the least healthy files and queuing the worst ones for repair.
+//
+// TODO / NOTE: Once we have upgraded the filesystem, we can replace this with
+// the tree-diving technique discussed in Sprint 5. For now we just iterate
+// through all of our in-memory files and chunks, and maintain a finite list of
+// the worst ones, and then iterate through it again when we need to find more
+// things to repair.
+func (r *Renter) threadedRepairScan() {
+	err := r.tg.Add()
+	if err != nil {
+		return
+	}
+	defer r.tg.Done()
+
+	for {
+		// Return if the renter has shut down.
+		select{
+		case <-r.tg.StopChan():
+			return
+		default:
+		}
+
+		// Grab the current set of contracts and a lookup table from file
+		// contract ids to host public keys. The lookup table has a mapping from
+		// all historic file contract ids, which is necessary when using the
+		// renter to perform downloads.
+		//
+		// TODO / NOTE: This code can be removed once files store the HostPubKey
+		// of the hosts they are using, instead of just the FileContractID.
+		currentContracts, fcidToHPK := r.managedCurrentContractsAndHistoricFCIDLookup()
```

Comment:
could this be split into two calls? I'm thinking along the lines of using `Contracts` to get `currentContracts`, and replacing the `fcidToHPK` map with a `contractor.ContractHostKey` method.